sklearn/cluster/_hierarchical.pyx

```diff
@@ -332,3 +332,80 @@ cdef class WeightedEdge:
                                               self.weight,
                                               self.a, self.b)
 
+
+################################################################################
+# Efficient labelling/conversion of MSTs to single linkage hierarchies
+
+cdef class UnionFind(object):
+
+    cdef np.ndarray parent_arr
+    cdef np.ndarray size_arr
+    cdef ITYPE_t next_label
+    cdef ITYPE_t *parent
+    cdef ITYPE_t *size
+
+    def __init__(self, N):
+        self.parent_arr = -1 * np.ones(2 * N - 1, dtype=ITYPE, order='C')
+        self.next_label = N
+        self.size_arr = np.hstack((np.ones(N, dtype=ITYPE),
+                                   np.zeros(N-1, dtype=ITYPE)))
+        self.parent = (<ITYPE_t *> self.parent_arr.data)
+        self.size = (<ITYPE_t *> self.size_arr.data)
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef void union(self, ITYPE_t m, ITYPE_t n):
+        self.size[self.next_label] = self.size[m] + self.size[n]
+        self.parent[m] = self.next_label
+        self.parent[n] = self.next_label
+        self.size[self.next_label] = self.size[m] + self.size[n]
+        self.next_label += 1
+
+        return
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef ITYPE_t fast_find(self, ITYPE_t n):
+        cdef ITYPE_t p
+        p = n
+        while self.parent_arr[n] != -1:
+            n = self.parent_arr[n]
+        # label up to the root
+        while self.parent_arr[p] != n:
+            p, self.parent_arr[p] = self.parent_arr[p], n
+        return n
+
+
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+cpdef np.ndarray[DTYPE_t, ndim=2] single_linkage_label(
```

Comment:
docstring?

Reply:
Should be fixed now. It is not extensive, but this is a fairly internal function.